src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Android/ChainPal.cs

```diff
@@ -35,7 +38,319 @@ public static bool ReleaseSafeX509ChainHandle(IntPtr handle)
+                X509Certificate2Collection? extraStore,
+                X509Certificate2Collection customTrustStore,
+                X509ChainTrustMode trustMode)
+            {
+                List<IntPtr> extraCerts = new List<IntPtr>() { cert.Handle };
+                if (extraStore != null)
+                {
+                    foreach (X509Certificate2 extraCert in extraStore)
+                    {
+                        extraCerts.Add(extraCert.Pal.Handle);
+                    }
+                }
+
+                List<IntPtr> customTrustCerts = new List<IntPtr>();
+                bool useCustomRootTrust = trustMode == X509ChainTrustMode.CustomRootTrust;
+                if (useCustomRootTrust && customTrustStore != null)
+                {
+                    foreach (X509Certificate2 custom in customTrustStore)
+                    {
+                        IntPtr certHandle = custom.Pal.Handle;
+                        if (custom.SubjectName.RawData.ContentsEqual(custom.IssuerName.RawData))
+                        {
+                            // Add self-issued certs to custom root trust cert
+                            customTrustCerts.Add(certHandle);
+                        }
+                        else
+                        {
+                            // Add non-self-issued certs to extra certs
+                            extraCerts.Add(certHandle);
+                        }
+                    }
+                }
+
+                IntPtr[] extraArray = extraCerts.ToArray();
+                _chainContext = Interop.AndroidCrypto.X509ChainCreateContext(
+                    ((AndroidCertificatePal)cert).SafeHandle,
+                    extraArray,
+                    extraArray.Length);
+
+                if (useCustomRootTrust)
+                {
+                    // Android does not support an empty set of trust anchors
+                    if (customTrustCerts.Count == 0)
+                    {
+                        throw new PlatformNotSupportedException(SR.Chain_EmptyCustomTrustNotSupported);
+                    }
+
+                    IntPtr[] customTrustArray = customTrustCerts.ToArray();
+                    int res = Interop.AndroidCrypto.X509ChainSetCustomTrustStore(_chainContext, customTrustArray, customTrustArray.Length);
+                    if (res != 1)
+                    {
+                        throw new CryptographicException();
+                    }
+                }
+            }
+
+            internal void Evaluate(
+                DateTime verificationTime,
+                OidCollection applicationPolicy,
+                OidCollection certificatePolicy,
+                X509RevocationMode revocationMode,
+                X509RevocationFlag revocationFlag)
+            {
+                Debug.Assert(_chainContext != null);
+
+                long timeInMsFromUnixEpoch = new DateTimeOffset(verificationTime).ToUnixTimeMilliseconds();
```

Comment:
I don't know which way Android revocation works, but would we do better with chain building if we always built the chain as of cert.NotBefore, then manually did the expired check ourselves afterwards?

Reply:
> I don't know which way Android revocation work

(Some libraries will return "not revoked" if the revocation time was in the future as compared to the verification time... some say "if it's on the list, it's revoked, I don't care about 'as of when'")

Reply:
Using the `cert.NotBefore` for building the initial chain should let us actually build a chain when the verification time is outside the cert's validity and get some better chain building.

Would there be situations where a chain should pass time validation (so we could build it on Android) with the specified verification time, but not with `cert.NotBefore` - e.g. an issuer cert had a later `NotBefore` that the verification time satisfies, but not `cert.NotBefore`? That seems weird to me, but I don't know the area and didn't find anything saying that the issuer cert's valid time should fully contain the issued cert's valid time. (I do see that .NET has the `NotTimeNested` flag that the doc says is deprecated.)

Reply:
Android revocation checking does care about 'as of when' and uses the verification time.

Reply:
> Android revocation checking does care about 'as of when' and uses the verification time.

😢 

So if we really cared to make expiration be soft-detected we'd need to do something like

* Check manually if it's expired
* Build with a valid time (e.g. the NotBefore time) and no revocation to get the full chain details
* For the newly chained-in certs, also manually check expiry (if the engine does nested validity time checking already then this can be skipped when the target cert wasn't expired).
* If revocation is requested
  * If expired: Slap RevocationStatusUnknown everywhere.
  * else: Use the requested time and build again with revocation on.

Sounds... complicated enough that I'm OK deferring it until complaints come in.